cirq/google/arg_func_langs.py

```diff
@@ -12,9 +12,196 @@
 
+import numpy as np
 import sympy
+from cirq.api.google import v2
 
-# Argument types for gates. ArgFunction's are not currently supported.
-ArgValue = Union[int, float, List[bool], str, sympy.Symbol]
+SUPPORTED_FUNCTIONS_FOR_LANGUAGE = {
+    '': set(),
+    'linear': {'add', 'mul'},
+    # None means any. Is used when inferring the language during serialization.
+    None: {'add', 'mul'},
+}
+
+# Argument types for gates.
+ARG_LIKE = Union[int, float, List[bool], str, sympy.Symbol, sympy.Add, sympy.
+                 Mul]
+
+LANGUAGE_ORDER = [
+    '',
+    'linear',
+]
+
+
+def _max_lang(langs: Iterable[str]) -> str:
+    i = max((LANGUAGE_ORDER.index(e) for e in langs), default=0)
+    return LANGUAGE_ORDER[i]
+
+
+def _infer_function_language_from_circuit(value: v2.program_pb2.Circuit) -> str:
+    return _max_lang({
+        e for moment in value.moments for op in moment.operations
+        for e in _function_languages_from_operation(op)
+    })
+
+
+def _infer_function_language_from_schedule(value: v2.program_pb2.Schedule
+                                          ) -> str:
+    return _max_lang({
+        e for op in value.scheduled_operations
+        for e in _function_languages_from_operation(op.operation)
+    })
+
+
+def _function_languages_from_operation(value: v2.program_pb2.Operation
+                                      ) -> Iterator[str]:
+    for arg in value.args.values():
+        yield from _function_languages_from_arg(arg)
+
+
+def _function_languages_from_arg(arg_proto: v2.program_pb2.Arg
+                                ) -> Iterator[str]:
+
+    which = arg_proto.WhichOneof('arg')
+    if which == 'func':
+        if arg_proto.func.type in ['add', 'mul']:
+            yield 'linear'
+            for a in arg_proto.func.args:
+                yield from _function_languages_from_arg(a)
+
+
+def _arg_to_proto(value: ARG_LIKE,
+                  *,
+                  arg_function_language: Optional[str],
+                  out: Optional[v2.program_pb2.Arg] = None
+                 ) -> v2.program_pb2.Arg:
+    """Writes an argument value into an Arg proto.
+
+    Args:
+        value: The value to encode.
+        arg_function_language: The language to use when encoding functions. If
+            this is set to None, it will be set to the minimal language
+            necessary to support the features that were actually used.
+        out: The proto to write the result into. Defaults to a new instance.
+
+    Returns:
+        The proto that was written into as well as the `arg_function_language`
+        that was used.
+    """
+
+    supported = SUPPORTED_FUNCTIONS_FOR_LANGUAGE.get(arg_function_language)
+    if supported is None:
+        raise ValueError(f'Unrecognized arg_function_language: '
+                         f'{arg_function_language!r}')
+
+    msg = v2.program_pb2.Arg() if out is None else out
+
+    def check_support(func_type: str) -> str:
+        if func_type not in cast(Set[str], supported):
```

Comment:
Would it be cleaner to put a type annotation on SUPPORTED_FUNCTIONS_FOR_LANGUAGE rather than need to cast here?  Will that work?

Reply:
The issue is actually that mypy things `supported` could be `None`. I rewrote the above lines to convince it otherwise. Also added the type annotation.